src/vulkan-renderer/world/collision.cpp

```diff
@@ -0,0 +1,143 @@
+        tmin = tymin;
+    }
+    if (tymax < tmax) {
+        tmax = tymax;
+    }
+
+    tzmin = (box_bounds[sign[2]].z - position.z) * inverse_dir.z;
+    tzmax = (box_bounds[1 - sign[2]].z - position.z) * inverse_dir.z;
+
+    if ((tmin > tzmax) || (tzmin > tmax)) {
+        return false;
+    }
+    if (tzmin > tmin) {
+        tmin = tzmin;
+    }
+    if (tzmax < tmax) {
+        tmax = tzmax;
+    }
+
+    return true;
+}
+
+void OctreeCollision::clear_hits() {
+    for (auto hit : m_hits) {
+        hit.reset();
+    }
+    m_hits_found = 0;
+}
+
+void OctreeCollision::add_hit(CollisionData collision) {
+    m_hits[m_hits_found] = collision;
+    m_hits_found++;
+}
+
+void OctreeCollision::sort_hits(const glm::vec3 position) {
+    // TODO: Refactor! This implementation is stupid. There must be a better way of doing this!
+    std::unordered_map<std::size_t, float> calculated_distances;
+    for (std::size_t i = 0; i < m_hits_found; i++) {
+        calculated_distances[i] = m_hits[i].squared_distance(position);
+        spdlog::trace("Collision {}, squared distance {}", i, calculated_distances[i]);
+    }
+
+    // TODO: Search for the lowest squared distance.
+    // TODO: Return sorted collisions.
+}
+
+std::optional<CollisionData> OctreeCollision::check_for_collision(const glm::vec3 position, const glm::vec3 direction) {
+    /// If the cube is empty, a collision with a ray is not possible,
+    /// and there are no sub cubes to check for collision either.
+    if (m_cube->type() == Cube::Type::EMPTY) {
+        return std::nullopt;
+    }
+
+    /// First, check if ray collides with bounding sphere.
+    /// This is much easier to calculate than collision with a bounding box.
+    if (ray_sphere_collision(position, direction)) {
+        /// Second, check if ray collides with bounding box.
+        if (ray_box_collision(m_cube->bounding_box(), position, direction)) {
+            if (m_cube->is_leaf()) {
+                /// This is a leaf node, a collision has been found.
+                /// TODO: Check which face and edge is selected.
+                return std::make_optional<CollisionData>(m_cube);
+            } else {
+                clear_hits();
+                /// Iterate through all sub cubes and check for collision.
+                const auto &subcubes = m_cube->childs();
+                for (std::int32_t i = 0; i < 8; i++) {
+                    if (subcubes[i]->type() != Cube::Type::EMPTY) {
+                        OctreeCollision subcollision(subcubes[i]);
+                        const auto collision_result = subcollision.check_for_collision(position, direction);
+                        if (collision_result) {
+                            spdlog::debug("Collided with subcube no {}.", i);
+                            add_hit(collision_result.value());
+                            // TODO: Check how many hits could be found at most (less than 8) and abort the loop.
+                        }
```

Comment:
Something along the lines of that:

```suggestion
            if (m_cube->is_leaf()) {
                /// This is a leaf node, a collision has been found.
                /// TODO: Check which face and edge is selected.
                
                add_hit(m_cube);
                // TODO: Check how many hits could be found at most (less than 8) and abort the loop.
                return std::make_optional<CollisionData>(m_cube);
            } else {
                clear_hits();
                /// Iterate through all sub cubes and check for collision.
                const auto &subcubes = m_cube->childs();
                for (std::int32_t i = 0; i < 8; i++) {
                    if (subcubes[i]->type() != Cube::Type::EMPTY) {
                        OctreeCollision subcollision(subcubes[i]);
                        subcollision.check_for_collision(position, direction);
```

Firstly, it feels (for me) more intuitive when the comment says "a collision has been found" that we add the hit directly afterwards.
Secondly, you wouldn't need to return anything in this method or "only" true/false if the you want to know how many and which cubes were hit.
The "root"-call of this method seems to return `std::nullopt` anyway (except for some rare cases) or do I miss something?

Reply:
It will return the final collision (if any found). Stay tuned for the final version of the PR code :)